test/extended/router/gatewayapicontroller.go

```diff
@@ -0,0 +1,539 @@
+		o.Expect(errCheck).NotTo(o.HaveOccurred(), "GatewayClass %q was not installed and accepted", gatewayClassName)
+		waitForIstioHealthy(oc)
+
+		g.By("Create Custom GatewayClass")
+		gatewayClass := buildGatewayClass(customGatewayClassName, gatewayClassControllerName)
+		gwc, err := gwapiClient.GatewayV1().GatewayClasses().Create(context.TODO(), gatewayClass, metav1.CreateOptions{})
+		if err != nil {
+			e2e.Logf("Gateway Class \"custom-gatewayclass\" already exists, or has failed to be created, checking its status")
+		}
+		errCheck = checkGatewayClass(oc, customGatewayClassName)
+		o.Expect(errCheck).NotTo(o.HaveOccurred(), "GatewayClass %q was not installed and accepted", gwc.Name)
+
+		g.By("Deleting Custom GatewayClass and confirming that it is no longer there")
+		err = gwapiClient.GatewayV1().GatewayClasses().Delete(context.Background(), customGatewayClassName, metav1.DeleteOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("check if default gatewayClass is accepted and ISTIO CR and pod are still available")
+		defaultCheck := checkGatewayClass(oc, gatewayClassName)
+		o.Expect(defaultCheck).NotTo(o.HaveOccurred())
+
+		g.By("Confirm that ISTIO CR is created and in healthy state")
+		resource := types.NamespacedName{Namespace: "openshift-ingress", Name: "openshift-gateway"}
+
+		istioStatus, errIstio := oc.AsAdmin().Run("get").Args("-n", resource.Namespace, "istio", resource.Name, "-o=jsonpath={.status.state}").Output()
+		o.Expect(errIstio).NotTo(o.HaveOccurred())
+		o.Expect(istioStatus).To(o.Equal(`Healthy`))
+	})
+
+	g.It("Ensure loadbalancer service and dnsRecord are created for a Gateway object", func() {
+		var lbAddress string
+		g.By("Ensure default GatewayClass is accepted")
+		errCheck := checkGatewayClass(oc, gatewayClassName)
+		o.Expect(errCheck).NotTo(o.HaveOccurred(), "GatewayClass %q was not installed and accepted", gatewayClassName)
+
+		g.By("Getting the default domain")
+		defaultIngressDomain, err := getDefaultIngressClusterDomainName(oc, time.Minute)
+		o.Expect(err).NotTo(o.HaveOccurred(), "failed to find default domain name")
+		defaultDomain := strings.Replace(defaultIngressDomain, "apps.", "gw-default.", 1)
+
+		g.By("Create the default gateway")
+		_, gwerr := createAndCheckGateway(oc, gatewayName, gatewayClassName, defaultDomain)
+		o.Expect(gwerr).NotTo(o.HaveOccurred(), "failed to create Gateway")
+
+		g.By("Verify the gateway's LoadBalancer service and DNSRecords")
+		// check gateway LB service, note that External-IP might be hostname (AWS) or IP (Azure/GCP)
+		lbService, err := oc.AdminKubeClient().CoreV1().Services("openshift-ingress").Get(context.Background(), gatewayName+"-openshift-default", metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		if lbService.Status.LoadBalancer.Ingress[0].Hostname != "" {
+			lbAddress = lbService.Status.LoadBalancer.Ingress[0].Hostname
+		} else {
+			lbAddress = lbService.Status.LoadBalancer.Ingress[0].IP
+		}
```

Comment:
The old logic is failing on GCP/Azure because it always gets empty string of `.Hostname`.